src/components/cylc/Graph.vue

```diff
@@ -0,0 +1,1644 @@
+        this,
+        QUERIES[queryName].replace('WORKFLOW_ID', this.workflowId)
+      )
+      if (!(queryName in this.subscriptions)) {
+        this.subscriptions[queryName] = {
+          id
+        }
+      }
+    },
+
+    unsubscribe (queryName) {
+      if (queryName in this.subscriptions) {
+        cytoscapeService.unsubscribe(
+          this.subscriptions[queryName].id
+        )
+      }
+    },
+
+    async workflowUpdated (workflows) {
+      try {
+        if (this.freeze === false) {
+          this.loading = true
+          const elements = {
+            nodes: [],
+            edges: []
+          }
+
+          if (!isUndefined(workflows)) {
+            each(workflows, (value, key) => {
+              each(value, (workflow, key) => {
+                if (!isEmpty(workflow.owner)) {
+                  this.workflowOwner = workflow.owner
+                }
+                if (!isEmpty(workflow.name)) {
+                  this.workflowName = workflow.name
+                }
+                if (has(workflow.nodesEdges, 'edges') && !isUndefined(workflow.nodesEdges.edges)) {
+                  elements.edges = this.getEdges(workflow.nodesEdges.edges)
+                }
+                if (has(workflow.nodesEdges, 'nodes') && !isUndefined(workflow.nodesEdges.nodes)) {
+                  elements.nodes = this.getNodes(workflow.nodesEdges.nodes)
+                }
+              })
+            })
+          }
+          console.debug('elements ==>>>> ', elements)
+          isEmpty(elements) ? console.warn('gdata is empty or undefined') : this.graphData = elements
+        }
+      } catch (error) {
+        console.error('workflowUpdated error: ', error)
+      }
+    },
+
+    getNodes (nodes) {
+      try {
+        const nodesArray = []
+        let nodeObj = {}
+        let todo = 0
+        let jobsArray = []
+        each(nodes, (node, key) => {
+          nodeObj = {
+            data: {
+              id: '',
+              name: '',
+              state: 'undefined',
+              parent: '',
+              label: '',
+              shape: 'ellipse',
+              runpercent: 0,
+              todo: 0,
+              running: 0,
+              cyclepoint: '',
+              submittedtime: '',
+              starttime: '',
+              finishtime: '',
+              jobs: '',
+              jobsarray: [],
+              status: ''
+            },
+            position: {
+
+            },
+            group: 'nodes',
+            removed: false,
+            selected: false,
+            selectable: true,
+            locked: false,
+            grabbable: true,
+            classes: ''
+          }
+          if (has(node, 'task') && !isEmpty(node.task) && has(node.task, 'name') && !isEmpty(node.task.name)) {
+            nodeObj.data.label = node.task.name
+          }
+          has(node, 'id') && !isEmpty(node.id) ? nodeObj.data.id = node.id : console.warn('workflowUpdated - node id is empty')
```

Comment:
Perhaps the tree component should also be more defensive. At the moment I'm trusting almost 100% on the GraphQL data to have some value, or at least be `undefined` or `null`. I think I had to check for the types returned in only a few places, but maybe the data returned is different. Just a comment, no action needed here :+1:

Reply:
The graph is very sensitive to bad data, hence the checks, I would like to use more functional filters like monads, probably overkill here...